src/morphodict/runserver/mobile_run_handler.py

```diff
@@ -0,0 +1,146 @@
+server’s listening socket. Python and/or Django have no idea that a listening
+socket can be closed out from underneath them, so the Django runserver listening
+code just sits there waiting for requests even though the socket is closed, and
+the app doesn’t work because the iOS web view can’t load anything from Django.
+
+The workaround, which is a bit hacky, is to poke around in the runserver
+internals, find that it calls a function called `run` imported from `basehttp`,
+which calls `TCPServer.serve_forever()`, and instead of calling that just once,
+call it in a loop which waits for a notification before starting the server back
+up on subsequent loop iterations.
+
+This reduces the iOS app’s interaction with the Django server code down to
+calling stop_server() when backgrounded, and resume_server() when brought to the
+foreground again.
+"""
+
+
+import socketserver
+from threading import Condition, Lock
+from typing import Optional
+
+from django.core.servers.basehttp import WSGIRequestHandler, WSGIServer
+
+
+class RunHandler:
+    def __init__(self, *, server_address, wsgi_handler, httpd_cls, ipv6, threading):
+        self.server_address = server_address
+        self.wsgi_handler = wsgi_handler
+        self.httpd_cls = httpd_cls
+        self.ipv6 = ipv6
+        self.threading = threading
+
+        self.httpd = None
+
+        self.condition = Condition()
+        self.httpd_lock = Lock()
+
+    def start(self):
+        while True:
+            ## Portions copied from Django’s basehttp.run
+            with self.httpd_lock:
+                assert self.httpd is None
+
+                self.httpd = self.httpd_cls(
+                    self.server_address, WSGIRequestHandler, ipv6=self.ipv6
+                )
+            if self.threading:
+                # ThreadingMixIn.daemon_threads indicates how threads will behave on an
+                # abrupt shutdown; like quitting the server by the user or restarting
+                # by the auto-reloader. True means the server will not wait for thread
+                # termination before it quits. This will make auto-reloader faster
+                # and will prevent the need to kill the server manually if a thread
+                # isn't terminating correctly.
+                self.httpd.daemon_threads = True
+            self.httpd.set_app(self.wsgi_handler)
+
+            print("calling serve_forever()")
+            # This method is somewhat misleadingly named: it will exit if
+            # `httpd.shutdown()` is called from a different thread.
+            self.httpd.serve_forever()
+            print("serve_forever stopped, waiting for notification to resume")
+
+            with self.condition:
+                self.condition.wait()
+                print("notified")
+
+    def resume(self):
+        print("RunHandler.resume")
+        with self.condition:
+            self.condition.notify()
+
+    def stop(self):
+        print("RunHandler.stop")
```

Comment:
Are these print statements routed through the logging function to show up in the device logs?

Reply:
Yup, python’s `sys.stdout` is rerouted so any `print` statements inside Python go to the system log. A hypothetical raw C `printf` inside some python extension wouldn’t be redirected because it wouldn’t involve the Python output layer at all, but I don’t think that’s a big deal.